src/protocol/attribution/mod.rs

```diff
@@ -3,28 +3,159 @@ pub mod aggregate_credit;
+    input::{
+        MCAggregateCreditOutputRow, MCApplyAttributionWindowInputRow,
+        MCCappedCreditsWithAggregationBit,
+    },
+};
 use crate::{
     error::Error,
-    ff::{Field, Gf2},
-    protocol::{basics::SecureMul, boolean::or::or, context::Context, BasicProtocols, RecordId},
+    ff::{Field, GaloisField, Gf2, PrimeField, Serializable},
+    helpers::query::IpaQueryConfig,
+    protocol::{
+        basics::SecureMul,
+        boolean::{bitwise_equal::bitwise_equal_gf2, or::or, RandomBits},
+        context::{Context, UpgradableContext, UpgradedContext, Validator},
+        ipa::IPAModulusConvertedInputRow,
+        modulus_conversion::{convert_bit, convert_bit_local, BitConversionTriple},
+        sort::generate_permutation::ShuffledPermutationWrapper,
+        step, BasicProtocols, RecordId,
+    },
     repeat64str,
     secret_sharing::{
-        replicated::semi_honest::AdditiveShare as Replicated, Linear as LinearSecretSharing,
+        replicated::{
+            malicious::{DowngradeMalicious, ExtendableField},
+            semi_honest::{AdditiveShare as Replicated, AdditiveShare as SemiHonestAdditiveShare},
+        },
+        Linear as LinearSecretSharing,
     },
-    seq_join::{assert_send, SeqJoin},
+    seq_join::assert_send,
 };
 use futures::future::try_join;
-use std::iter::{empty, zip};
+use std::iter::{empty, once, zip};
 
-use super::{
-    boolean::bitwise_equal::bitwise_equal_gf2,
-    context::SemiHonestContext,
-    modulus_conversion::{convert_bit, convert_bit_local, BitConversionTriple},
-};
+/// Performs a set of attribution protocols on the sorted IPA input.
+///
+/// # Errors
+/// propagates errors from multiplications
+pub async fn secure_attribution<C, S, SB, F, BK>(
+    sh_ctx: C,
+    validator: C::Validator<F>,
+    binary_validator: C::Validator<Gf2>,
+    sorted_match_keys: Vec<Vec<SB>>,
+    sorted_rows: Vec<IPAModulusConvertedInputRow<F, S>>,
+    config: IpaQueryConfig,
+) -> Result<Vec<MCAggregateCreditOutputRow<F, SemiHonestAdditiveShare<F>, BK>>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S> + RandomBits<F, Share = S>,
+    S: LinearSecretSharing<F> + BasicProtocols<C::UpgradedContext<F>, F> + Serializable + 'static,
+    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = SB> + Context,
+    SB: LinearSecretSharing<Gf2> + BasicProtocols<C::UpgradedContext<Gf2>, Gf2> + 'static,
+    Vec<SB>: DowngradeMalicious<Target = Vec<SemiHonestAdditiveShare<Gf2>>>,
+    F: PrimeField + ExtendableField,
+    BK: GaloisField,
+    ShuffledPermutationWrapper<S, C::UpgradedContext<F>>: DowngradeMalicious<Target = Vec<u32>>,
+    MCCappedCreditsWithAggregationBit<F, S>: DowngradeMalicious<
+        Target = MCCappedCreditsWithAggregationBit<F, SemiHonestAdditiveShare<F>>,
+    >,
+    MCAggregateCreditOutputRow<F, S, BK>:
+        DowngradeMalicious<Target = MCAggregateCreditOutputRow<F, SemiHonestAdditiveShare<F>, BK>>,
```

Comment:
The `Target = ` bound on `DowngradeMalicious` turns out to be essential (otherwise, when we try to downgrade, the compiler observes that we could be producing anything.

I did consider adding more letters to the signature, but it's not that much of a gain.  More associated types would help, but I found myself blocked by the lack of defaults for associated types in the language.